File: client-new/src/api/gamification.ts
import api from './client';

/* ============ Types ============ */

export interface CoinBalance {
  balance: number;
  total_earned: number;
  total_spent: number;
}

export interface CoinTransaction {
  id: string;
  amount: number;
  tx_type: string;
  reason: string | null;
  granted_by_name: string | null;
  created_at: string;
}

export interface ShopItem {
  id: string;
  title: string;
  description: string | null;
  price: number;
  icon: string | null;
  category: string;
  image_url: string | null;
  is_active: boolean;
  stock: number | null;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  level_required: number;
  is_featured: boolean;
  created_at: string;
}

export interface ShopShowcaseItem extends ShopItem {
  owned_count: number;
  can_buy: boolean;
  is_locked: boolean;
  next_tier_required: number;
}

export interface ShopPurchase {
  id: string;
  item_id: string;
  item_title: string;
  price_paid: number;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  category: 'status' | 'badge' | 'perk';
  created_at: string;
}

export interface EquippedItem {
  category: string;
  purchase_id: string;
  item_id: string;
  item_title: string;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  equipped_at: string;
}

export interface Achievement {
  id: string;
  title: string;
  category: string;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  level: number;
  progress: number;
  target: number;
  unlocked: boolean;
  icon: string | null;
}

export interface TestResult {
  id: string;
  topic_id: string;
  score: number;
  total: number;
  passed: boolean;
  attempt: number;
  created_at: string;
}

export interface UserKPI {
  user_id: string;
  user_name: string;
  avatar_url: string | null;
  total_time_minutes: number;
  topics_completed: number;
  topics_total: number;
  tests_passed: number;
  tests_total: number;
  avg_test_score: number;
  coins_balance: number;
  completion_pct: number;
  speed_topics_per_day: number;
  retention_pct: number;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  user_name: string;
  avatar_url: string | null;
  coins_balance: number;
  topics_completed: number;
  avg_test_score: number;
  total_time_hours: number;
  tasks_assigned: number;
  tasks_completed_week: number;
  tasks_completed_day: number;
  tasks_completed_month: number;
  tasks_completed_year: number;
  tasks_overdue: number;
  training_progress_pct: number;
  coins_earned_tasks: number;
  anti_cheat_score: number;
  anti_cheat_flags: string[];
}

export interface SectionAccess {
  user_id: string;
  section_keys: string[];
}

/* ============ API ============ */

export const gamificationApi = {
  // Coins
  getBalance: () =>
    api.get<CoinBalance>('/gamification/coins/balance'),
  getCoinHistory: () =>
    api.get<CoinTransaction[]>('/gamification/coins/history'),
  grantCoins: (data: { user_id: string; amount: number; reason: string }) =>
    api.post<CoinTransaction>('/gamification/coins/grant', data),

  // Shop
  getShopItems: () =>
    api.get<ShopItem[]>('/gamification/shop/items'),
  getShopShowcase: () =>
    api.get<ShopShowcaseItem[]>('/gamification/shop/showcase'),
  createShopItem: (data: { title: string; description?: string; price: number; icon?: string; category?: string; image_url?: string; stock?: number; rarity?: string; level_required?: number; is_featured?: boolean }) =>
    api.post<ShopItem>('/gamification/shop/items', data),
  updateShopItem: (id: string, data: Partial<ShopItem>) =>
    api.put<ShopItem>(`/gamification/shop/items/${encodeURIComponent(id)}`, data),
  deleteShopItem: (id: string) =>
    api.delete(`/gamification/shop/items/${encodeURIComponent(id)}`),
  buyItem: (id: string) =>
    api.post<ShopPurchase>(`/gamification/shop/buy/${encodeURIComponent(id)}`),
  getMyPurchases: () =>
    api.get<ShopPurchase[]>('/gamification/shop/my-purchases'),
  getAchievements: () =>
    api.get<Achievement[]>('/gamification/shop/achievements'),
  getEquipped: () =>
    api.get<EquippedItem[]>('/gamification/shop/equipped'),
  equipPurchase: (purchaseId: string) =>
    api.post<EquippedItem>(`/gamification/shop/equip/${encodeURIComponent(purchaseId)}`),

  // Tests
  submitTest: (data: { topic_id: string; score: number; total: number }) =>
    api.post<TestResult>('/gamification/test/submit', data),
  getTestResults: () =>
    api.get<TestResult[]>('/gamification/test/results'),
  getTopicTestResult: (topicId: string) =>
    api.get<TestResult>(`/gamification/test/topic/${encodeURIComponent(topicId)}`),

  // KPI
  getMyKPI: () =>
    api.get<UserKPI>('/gamification/kpi/me'),
  getUserKPI: (userId: string) =>
    api.get<UserKPI>(`/gamification/kpi/user/${encodeURIComponent(userId)}`),

  // Leaderboard
  getLeaderboard: () =>
    api.get<LeaderboardEntry[]>('/gamification/leaderboard'),

  // session/end при уходе со страницы — 401 не должен запускать refresh и редирект (шторм в консоли)
  sessionPing: () =>
    api.post('/gamification/session/ping'),
  sessionEnd: () =>
    api.post('/gamification/session/end', null, { _silent401: true } as any),

  // Section access
  grantAccess: (data: { user_id: string; section_keys: string[] }) =>
    api.post('/gamification/access/grant', data),
  setAccess: (userId: string, sectionKeys: string[]) =>
    api.put(`/gamification/access/${encodeURIComponent(userId)}`, { section_keys: sectionKeys }),
  getAccess: (userId: string) =>
    api.get<SectionAccess>(`/gamification/access/${encodeURIComponent(userId)}`),

  // KPI Manager & Drops
  getActiveDrops: () =>
    api.get<KPIDrop[]>('/gamification/kpi/drops/active'),
  submitPerformanceReview: (data: PerformanceReviewCreate) =>
    api.post<PerformanceReview>('/gamification/kpi/reviews', data),
  getManagerKPIDetails: () =>
    api.get<ManagerKPIDetails>('/gamification/kpi/manager/details'),
  simulateKPIDrop: (data: { kpi_type: string; drop_value: number; employee_id?: string }) =>
    api.post<any>(`/gamification/kpi/drops/simulate?kpi_type=${encodeURIComponent(data.kpi_type)}&drop_value=${data.drop_value}${data.employee_id ? `&employee_id=${encodeURIComponent(data.employee_id)}` : ''}`),
  
  // Admin dashboard
  getDepartmentKPIHealth: () =>
    api.get<DepartmentKPIHealth[]>('/gamification/admin/department-kpi-health'),
  getManagerReactivity: () =>
    api.get<ManagerReactivity[]>('/gamification/admin/manager-reactivity'),
};

export interface KPIDrop {
  id: string;
  employee_id: string;
  employee_name: string | null;
  kpi_type: string;
  drop_value: number;
  drop_date: string;
  resolved: boolean;
  notification_sent: boolean;
}

export interface PerformanceReviewCreate {
  drop_id?: string;
  kpi_type: string;
  reason: string;
  action: string;
  comment?: string;
}

export interface PerformanceReview {
  id: string;
  drop_id: string | null;
  manager_id: string;
  manager_name: string | null;
  review_date: string;
  kpi_type: string;
  reason: string;
  action: string;
  comment: string | null;
  reaction_days: number | null;
  is_overtime: boolean;
  created_at: string;
}

export interface ManagerKPIDetails {
  manager_id: string;
  current_kpi2: number | null;
  reviews_count: number;
  total_days: number;
  overtime_reviews_count: number;
  total_overtime_percent: number;
  active_drops: KPIDrop[];
  recent_reviews: PerformanceReview[];
}

export interface DepartmentKPIHealth {
  department_id: string | null;
  employee_count: number;
  avg_kpi1_deadlines: number | null;
  avg_kpi2_punctuality: number | null;
  avg_kpi3_initiative: number | null;
  avg_kpi4_overtime: number | null;
  avg_kpi5_quality: number | null;
  avg_kpi8_attentiveness: number | null;
  avg_kpi9_bonus: number | null;
  avg_kpi10_responsibility: number | null;
}

export interface ManagerReactivity {
  manager_id: string;
  manager_name: string;
  active_drops_count: number;
  conducted_reviews_count: number;
  avg_reaction_days: number | null;
  manager_kpi1_reaction_index: number | null;
  manager_kpi3_responsibility: number | null;
  manager_kpi4_attentiveness: number | null;
}

File: server/app/schemas/gamification.py
# Pydantic-схемы геймификации
import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


# --- Coins ---
class CoinBalanceOut(BaseModel):
    balance: float = 0
    total_earned: float = 0
    total_spent: float = 0


class CoinTransactionOut(BaseModel):
    id: uuid.UUID
    amount: float
    tx_type: str
    reason: Optional[str] = None
    granted_by_name: Optional[str] = None
    created_at: datetime
    class Config:
        from_attributes = True


class AdminCoinGrant(BaseModel):
    user_id: uuid.UUID
    amount: float = Field(..., gt=0, le=10000)
    reason: str = Field(..., min_length=3, max_length=500)


# --- Shop ---
class ShopItemCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    price: int = Field(..., gt=0)
    icon: Optional[str] = None
    category: str = Field(default="status", pattern="^(status|badge|perk)$")
    image_url: Optional[str] = None
    stock: Optional[int] = None
    rarity: str = Field(default="common", pattern="^(common|rare|epic|legendary)$")
    level_required: int = Field(default=1, ge=1, le=100)
    is_featured: bool = False


class ShopItemUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    price: Optional[int] = Field(None, gt=0)
    icon: Optional[str] = None
    category: Optional[str] = None
    image_url: Optional[str] = None
    stock: Optional[int] = None
    is_active: Optional[bool] = None
    rarity: Optional[str] = None
    level_required: Optional[int] = None
    is_featured: Optional[bool] = None


class ShopItemOut(BaseModel):
    id: uuid.UUID
    title: str
    description: Optional[str] = None
    price: int
    icon: Optional[str] = None
    category: str
    image_url: Optional[str] = None
    is_active: bool
    stock: Optional[int] = None
    rarity: str = "common"
    level_required: int = 1
    is_featured: bool = False
    created_at: datetime
    class Config:
        from_attributes = True


class ShopPurchaseOut(BaseModel):
    id: uuid.UUID
    item_id: uuid.UUID
    item_title: str = ""
    price_paid: float
    rarity: str = "common"
    category: str = "status"
    created_at: datetime
    class Config:
        from_attributes = True


class ShopShowcaseItemOut(ShopItemOut):
    owned_count: int = 0
    can_buy: bool = True
    is_locked: bool = False
    next_tier_required: int = 1


class EquipRequest(BaseModel):
    purchase_id: uuid.UUID


class EquippedItemOut(BaseModel):
    category: str
    purchase_id: uuid.UUID
    item_id: uuid.UUID
    item_title: str
    rarity: str = "common"
    equipped_at: datetime


class AchievementOut(BaseModel):
    id: str
    title: str
    category: str
    rarity: str
    level: int
    progress: int
    target: int
    unlocked: bool
    icon: Optional[str] = None


# --- Test Results ---
class TestSubmit(BaseModel):
    topic_id: uuid.UUID
    score: int = Field(..., ge=0)
    total: int = Field(..., gt=0)


class TestResultOut(BaseModel):
    id: uuid.UUID
    topic_id: uuid.UUID
    score: int
    total: int
    passed: bool
    attempt: int
    created_at: datetime
    class Config:
        from_attributes = True


# --- KPI ---
class UserKPIOut(BaseModel):
    user_id: uuid.UUID
    user_name: str
    avatar_url: Optional[str] = None
    total_time_minutes: int = 0
    topics_completed: int = 0
    topics_total: int = 0
    tests_passed: int = 0
    tests_total: int = 0
    avg_test_score: float = 0.0
    coins_balance: float = 0
    completion_pct: float = 0.0
    speed_topics_per_day: float = 0.0  # Скорость прохождения
    retention_pct: float = 0.0  # Усвоение материала (avg тест %)
    
    # Employee 17-point KPI system fields
    kpi1_deadlines: Optional[float] = None
    kpi2_punctuality: Optional[float] = None
    kpi3_initiative: Optional[float] = None
    kpi4_overtime: Optional[float] = None
    kpi5_quality: Optional[float] = None
    kpi8_attentiveness: Optional[float] = None
    kpi9_bonus: Optional[float] = None
    kpi10_responsibility: Optional[float] = None
    
    # Manager-specific KPI fields
    manager_kpi1_reaction_index: Optional[float] = None
    manager_kpi2_reaction_days: Optional[float] = None
    manager_kpi3_responsibility: Optional[float] = None
    manager_kpi4_attentiveness: Optional[float] = None


class SessionPing(BaseModel):
    """Heartbeat для отслеживания времени на платформе"""
    pass


class LeaderboardEntry(BaseModel):
    rank: int
    user_id: uuid.UUID
    user_name: str
    avatar_url: Optional[str] = None
    coins_balance: float = 0
    topics_completed: int = 0
    avg_test_score: float = 0.0
    total_time_hours: float = 0.0
    # Detailed task stats
    tasks_assigned: int = 0
    tasks_completed_week: int = 0
    tasks_completed_day: int = 0
    tasks_completed_month: int = 0
    tasks_completed_year: int = 0
    tasks_overdue: int = 0
    training_progress_pct: float = 0.0
    coins_earned_tasks: float = 0
    anti_cheat_score: int = 100
    anti_cheat_flags: list[str] = Field(default_factory=list)


# --- Section Access ---
class SectionAccessGrant(BaseModel):
    user_id: uuid.UUID
    section_keys: list[str] = Field(..., min_length=1)


class UserSectionAccessOut(BaseModel):
    user_id: uuid.UUID
    section_keys: list[str] = Field(default_factory=list)


# --- KPI Drops & Performance Reviews ---
class KPIDropOut(BaseModel):
    id: uuid.UUID
    employee_id: uuid.UUID
    employee_name: Optional[str] = None
    kpi_type: str
    drop_value: float
    drop_date: datetime
    resolved: bool
    notification_sent: bool
    class Config:
        from_attributes = True


class PerformanceReviewCreate(BaseModel):
    drop_id: Optional[uuid.UUID] = None
    kpi_type: str
    reason: str
    action: str
    comment: Optional[str] = None


class PerformanceReviewOut(BaseModel):
    id: uuid.UUID
    drop_id: Optional[uuid.UUID] = None
    manager_id: uuid.UUID
    manager_name: Optional[str] = None
    review_date: datetime
    kpi_type: str
    reason: str
    action: str
    comment: Optional[str] = None
    reaction_days: Optional[float] = None
    is_overtime: bool
    created_at: datetime
    class Config:
        from_attributes = True


class ManagerKPIDetailsOut(BaseModel):
    manager_id: uuid.UUID
    current_kpi2: Optional[float] = None
    reviews_count: int = 0
    total_days: float = 0.0
    overtime_reviews_count: int = 0
    total_overtime_percent: int = 0
    active_drops: list[KPIDropOut] = Field(default_factory=list)
    recent_reviews: list[PerformanceReviewOut] = Field(default_factory=list)


# --- Admin Dashboard ---
class DepartmentKPIHealthOut(BaseModel):
    department_id: Optional[str] = None
    employee_count: int = 0
    avg_kpi1_deadlines: Optional[float] = None
    avg_kpi2_punctuality: Optional[float] = None
    avg_kpi3_initiative: Optional[float] = None
    avg_kpi4_overtime: Optional[float] = None
    avg_kpi5_quality: Optional[float] = None
    avg_kpi8_attentiveness: Optional[float] = None
    avg_kpi9_bonus: Optional[float] = None
    avg_kpi10_responsibility: Optional[float] = None


class ManagerReactivityOut(BaseModel):
    manager_id: uuid.UUID
    manager_name: str
    active_drops_count: int = 0
    conducted_reviews_count: int = 0
    avg_reaction_days: Optional[float] = None
    manager_kpi1_reaction_index: Optional[float] = None
    manager_kpi3_responsibility: Optional[float] = None
    manager_kpi4_attentiveness: Optional[float] = None
